common/futures-semaphore/src/lib.rs

```diff
@@ -72,82 +68,47 @@ impl Semaphore {
         self.available_permits() == 0
     }
 
-    /// Acquire an available permit from the semaphore, blocking until none are
-    /// available.
+    /// Acquire an available permit from the semaphore.
+    ///
+    /// If there are no permits currently available, the future will wait until
+    /// a permit is released back to the semaphore.
     pub async fn acquire(&self) -> Permit {
-        let permit = Permit {
+        // Acquire a permit and then immediately "forget" it (drop it without
+        // returning the permit to the pool of available permits). We will
+        // manually add a new permit when our `Permit` RAII guard is dropped.
+        //
+        // We do this instead of holding tokio's `SemaphorePermit` in the RAII
+        // guard due to lifetime issues--tokio's `SemaphorePermit` needs a
+        // direct reference to the `Semaphore`; however, we want to send our
+        // `Permit` to other tasks or threads, so we need an `Arc` to the
+        // semaphore so the permit doesn't outlive the semaphore.
```

Comment:
I'm a little worried about this, though you've implemented what you've described here. Do we practically need to "send" the permit to other threads?

Reply:
If we make the restriction that there can only be one handle for the `BoundedExecutor`, then I think the following would work: 

```rust
/// The wrapped tokio [`Semaphore`](TokioSemaphore) and total permit capacity.
#[derive(Debug)]
struct Inner {
    waker: AtomicWaker,
    active_permits: AtomicUsize,
    capacity: usize,
}

/// A futures-aware semaphore.
#[derive(Debug)]
pub struct Semaphore {
    inner: Arc<Inner>,
}

/// A permit acquired from a semaphore, allowing access to a shared resource.
/// Dropping a `Permit` will release it back to the semaphore.
#[derive(Debug)]
pub struct Permit {
    inner: Arc<Inner>,
}

impl Semaphore {
    /// Create a new semaphore with `capacity` number of available permits.
    pub fn new(capacity: usize) -> Self {
        Self {
            inner: Arc::new(Inner {
                waker: AtomicWaker::new(),
                active_permits: AtomicUsize::new(0),
                capacity,
            }),
        }
    }

    /// Acquire an available permit from the semaphore.
    ///
    /// If there are no permits currently available, the future will wait until
    /// a permit is released back to the semaphore.
    pub async fn acquire(&self) -> Permit {
        future::poll_fn(|cx| self.poll_acquire(cx)).await
    }

    fn poll_acquire(&self, cx: &mut Context<'_>) -> Poll<Permit> {
        self.inner.waker.register(cx.waker());
        let active_permits = self.inner.active_permits.load(Ordering::SeqCst);

        if active_permits == self.inner.capacity {
            Poll::Pending
        } else {
            self.inner.active_permits.fetch_add(1, Ordering::SeqCst);
            Poll::Ready(Permit {
                inner: Arc::clone(&self.inner),
            })
        }
    }
}

impl Drop for Permit {
    fn drop(&mut self) {
        self.inner.active_permits.fetch_sub(1, Ordering::SeqCst);
        self.inner.waker.wake();
    }
}
```

Though I would want to write some more exhaustive [tokio/loom](https://github.com/tokio-rs/loom) concurrency tests.